File: scalability/corvid_helics_lib/data_federates/query-federate.cpp
#include <boost/system/error_code.hpp>
#include <helics/application_api/MessageFederate.hpp>
#include <helics/application_api/Publications.hpp>
#include <helics/application_api/Inputs.hpp>
#include <helics/core/CoreTypes.hpp>

#include <iostream>
#include <sstream>
#include <string>
#include <fstream>
#include <chrono>
#include <thread>
#include <exception>
#include <cstdlib>
#include <filesystem>

#include <boost/optional.hpp>
#include <boost/json.hpp>

#include "stopwatch.hpp"

#include "query_federate_input.hpp"
#include "websocket_client.hpp"
#include "json_templates.hpp"
#include "local_log_helper.hpp"

namespace
{

std::optional<data::QueryFederateInput> GetQueryFederateInput(int argc, char **argv)
{
    std::optional<data::QueryFederateInput> query_input{};

    if (argc < 2)
    {
        std::cerr << "Missing JSON: No json file provided in execution call.\n";
        return query_input;
    }

    const std::filesystem::path cwd = std::filesystem::current_path();
    const std::string json_file = std::string(argv[1]);
    const std::filesystem::path json_path = json_file;
    if (!std::filesystem::exists(json_path))
    {
        std::cerr << "Missing JSON: " << json_path << " (expected in " << cwd << ")\n";
        return query_input;
    }

    query_input = utils::FromJsonFile<data::QueryFederateInput>(json_file);
    return query_input;
}

helics::MessageFederate GetFederate(const data::QueryFederateInput &config, utils::LocalLogHelper &log)
{
    helics::FederateInfo fi;
    fi.loadInfoFromJson(config.fed_info_json);

    helics::MessageFederate msg_fed(config.federate_name, fi);
    log << "Message Federate created successfully.\n";

    return msg_fed;
}

std::string DebugTimeQueryLoop(double granted_time, helics::MessageFederate &msg_fed)
{
    utils::Stopwatch loop_watch;
    loop_watch.Start();

    std::stringstream ss;
    ss << "\n##########################################\n";
    ss << "Granted Time: " << granted_time << "\n";
    ss << msg_fed.query("root", "global_time_debugging") << "\n";
    ss << "Query Execution Time: " << loop_watch.ElapsedMilliseconds() << " ms\n";
    ss << "##########################################\n";

    return ss.str();
}

std::string DiscreteQueriesLoop(double granted_time, helics::MessageFederate &msg_fed)
{
    utils::Stopwatch loop_watch;
    loop_watch.Start();

    std::stringstream ss;
    ss << "\n##########################################\n";
    ss << "Granted Time: " << granted_time << "\n";
    ss << "Name: " << msg_fed.query("root", "name") << "\n";
    ss << "Address: " << msg_fed.query("root", "address") << "\n";
    ss << "IsInit: " << msg_fed.query("root", "isinit") << "\n";
    ss << "IsConnected: " << msg_fed.query("root", "isconnected") << "\n";
    ss << "Query Execution Time: " << loop_watch.ElapsedMilliseconds() << " ms\n";
    ss << "##########################################\n";

    return ss.str();
}

double PerformLoop(helics::MessageFederate &msg_fed, const double total_time, const double period,
                   utils::LocalLogHelper &log)
{
    utils::Stopwatch main_watch;

    double granted_time = 0.0;

    main_watch.Start();
    while (granted_time + period <= total_time)
    {
        granted_time = msg_fed.requestTime(granted_time + period);

        // std::string output_string = DebugTimeQueryLoop(granted_time, msg_fed);
        std::string output_string = DiscreteQueriesLoop(granted_time, msg_fed);

        log << output_string;
    }
    double main_loop_ms = main_watch.ElapsedMilliseconds();

    std::stringstream end;
    end << "\n##########################################\n"
        << "Total Loop Time: " << main_loop_ms << " ms"
        << "\n##########################################\n"
        << "\nFederate finalized.\nGranted time: " << granted_time << "\n";
    log << end.str();

    return granted_time;
}

double ExecuteFederate(const data::QueryFederateInput &config, utils::LocalLogHelper &log)
{
    helics::MessageFederate msg_fed = GetFederate(config, log);
    const double period = msg_fed.getTimeProperty(HELICS_PROPERTY_TIME_PERIOD);

    double granted_time = -1.0;

    try
    {
        msg_fed.enterExecutingMode();

        // Sleep for a few seconds to enure the cosim is fully setup (this is the recommended approach....booo)
        std::this_thread::sleep_for(std::chrono::seconds(5));

        granted_time = PerformLoop(msg_fed, config.total_time, period, log);
    }
    catch (const std::exception &e)
    {
        log << "Error: " << e.what() << std::endl;
    }

    // Remember to finalize the fed
    msg_fed.finalize();

    return granted_time;
}
} // namespace

int main(int argc, char **argv)
{
    int ret_val = EXIT_FAILURE;
    auto client = std::make_shared<utils::WebSocketClient>();

    try
    {
        const std::optional<data::QueryFederateInput> query_input = GetQueryFederateInput(argc, argv);
        if (!query_input)
        {
            return EXIT_FAILURE;
        }

        utils::LocalLogHelper log(query_input.value().local_log_file);
        if (!log.IsOpen())
        {
            std::cerr << "Could not open file '" << query_input.value().local_log_file << "'!" << std::endl;
            return EXIT_FAILURE;
        }

        // Configure the client
        client->SetOnMessage([&log](const std::string &msg) { log << "Received: " << msg << std::endl; });
        client->SetOnError([&log](const boost::system::error_code &ec, const std::string &what)
                           { log << what << ": " << ec.message() << std::endl; });

        client->AsyncRun();
        client->Connect(query_input.value().client_details.host, query_input.value().client_details.port,
                        query_input.value().client_details.target,
                        [&log](const boost::system::error_code &ec)
                        {
                            if (ec)
                            {
                                log << "Error Connecting: " << ec.message() << "\n";
                            }
                            else
                            {
                                log << "Connected!\n";
                            }
                        });

        log.SetOnWriteCallback([&client](const std::string &msg) { client->Send(msg); });

        // Configure and launch the federate
        const double granted_time = ExecuteFederate(query_input.value(), log);
        if (granted_time < 0.0)
        {
            log << "Could not perform simulation! Federate finalized.\nGranted time: " << granted_time;
        }

        ret_val = EXIT_SUCCESS;
    }
    catch (const std::exception &e)
    {
        std::cerr << "Fatal error: " << e.what() << std::endl;
    }
    catch (...)
    {
        std::cerr << "Unknown Exception: An object that does not inherit std::exception was thrown!" << std::endl;
    }

    client->CloseConnection();
    client->StopRun();

    return ret_val;
}

File: scalability/corvid_helics_lib/data_federates/query-federate.py
import helics

import time
import os
import json
import argparse

from dataclasses import dataclass

@dataclass
class SubscriptionDetails:
    name: str
    units: str

class Stopwatch:
    def __init__(self):
        self.start_time = None

    def start(self):
        self.start_time = time.perf_counter()

    def get_elapsed_ms(self):
        current = time.perf_counter()
        elapsed = current - self.start_time
        return elapsed * 1000

def get_helics_federate(config):
    federate_name = config["name"]
    core_type = "zmq"
    if "core_type" in config.keys():
        core_type = config["core_type"]
    core_init_string = "--federates=1"
    if "core_init_string" in config.keys():
        core_init_string = config["core_init_string"]
    time_property = 1.0
    if "time_property" in config.keys():
        time_property = config["time_property"]

    fed_info = helics.helicsCreateFederateInfo()

    helics.helicsFederateInfoSetCoreName(fed_info, federate_name + "_core")
    helics.helicsFederateInfoSetCoreTypeFromString(fed_info, core_type)
    helics.helicsFederateInfoSetCoreInitString(fed_info, core_init_string)
    helics.helicsFederateInfoSetTimeProperty(fed_info, helics.helics_property_time_delta, time_property)

    return helics.helicsCreateMessageFederate(federate_name, fed_info)

def debug_time_query_loop(granted_time, federate):
    loop_watch = Stopwatch()

    output_str = []
    output_str.append("\n###################\n")
    output_str.append(f"Granted Time: {granted_time}\n")

    loop_watch.start()

    debugging_time_query = helics.helicsCreateQuery("root", "global_time_debugging")
    debugging_time_value = helics.helicsQueryExecute(debugging_time_query, federate)
    output_str.append(json.dumps(debugging_time_value))

    loop_execution_ms = loop_watch.get_elapsed_ms()
    output_str.append(f"\nQuery Execution Time: {loop_execution_ms} ms\n")
    output_str.append("###################\n")

    return "".join(output_str)


def discrete_loops(granted_time, federate):
    loop_watch = Stopwatch()

    output_str = []
    output_str.append("\n###################\n")
    output_str.append(f"Granted Time: {granted_time}\n")

    loop_watch.start()

    name_query = helics.helicsCreateQuery("root", "name")
    name_value = helics.helicsQueryExecute(name_query, federate)
    output_str.append(f"Name: {name_value}\n")

    address_query = helics.helicsCreateQuery("root", "address")
    address_value = helics.helicsQueryExecute(address_query, federate)
    output_str.append(f"Address: {address_value}\n")

    is_init_query = helics.helicsCreateQuery("root", "isinit")
    is_init_value = helics.helicsQueryExecute(is_init_query, federate)
    output_str.append(f"IsInit: {is_init_value}\n")

    is_connected_query = helics.helicsCreateQuery("root", "isconnected")
    is_connected_value = helics.helicsQueryExecute(is_connected_query, federate)
    output_str.append(f"IsConnected: {is_connected_value}\n")

    loop_execution_ms = loop_watch.get_elapsed_ms()
    output_str.append(f"Query Execution Time: {loop_execution_ms} ms\n")
    output_str.append("###################\n")

    return "".join(output_str)


def perform_loop(federate, total_time, period):
    with open("query-federate.log", "w") as output_file:
        main_watch = Stopwatch()
        granted_time = 0.0

        main_watch.start()
        while granted_time + period <= total_time:
            granted_time = helics.helicsFederateRequestTime(federate, granted_time + period)

            # output_string = debug_time_query_loop(granted_time, federate)
            output_string = discrete_loops(granted_time, federate)

            output_file.write(output_string)

        main_loop_ms = main_watch.get_elapsed_ms()
        output_str = []
        output_str.append("\n###################\n")
        output_str.append(f"Total Loop Time: {main_loop_ms} ms\n")
        output_str.append("\n###################\n")
        info_string = "".join(output_str)
        output_file.write(info_string)

def execute_federate(json_input):
    if json_input is None or not isinstance(json_input, str):
        return

    path_to_config = os.path.abspath(json_input)
    if not os.path.exists(path_to_config):
        print("Unable to find file '{}'".format(path_to_config))
        print("Returning...")
        return

    with open(path_to_config) as json_file:
        config = json.loads(json_file.read())

    federate = get_helics_federate(config)

    total_time = config["total_time"]
    period = 1.0
    if "period" in config.keys():
        period = config["period"]

    # Sleep for a few seconds to ensure the cosim is fully setup (this is recommended...booo)
    time.sleep(5)

    try:
        helics.helicsFederateEnterExecutingMode(federate)
        perform_loop(federate, total_time, period)
    finally:
        helics.helicsFederateFinalize(federate)
        helics.helicsCloseLibrary()

def main():
    parser = argparse.ArgumentParser(description="A federate to run queries")
    parser.add_argument("json_config_file", help="Path to the json configuration file")

    args = parser.parse_args()

    execute_federate(args.json_config_file)

if __name__ == "__main__":
    main()